agent/system_metrics.go

```diff
@@ -0,0 +1,245 @@
+	interval    time.Duration
+	collectors  []MetricCollector
+	id          string
+	client      *system_metrics.SystemMetricsClient
+	errored     bool
+	closed      bool
+}
+
+func NewSystemMetricsCollector(ctx context.Context, logger client.LoggerProducer, interval time.Duration,
+	c client.Communicator, t *task.Task, collectors []MetricCollector) (*SystemMetricsCollector, error) {
+
+	if interval < 0 {
+		return nil, errors.New("interval cannot be negative")
+	}
+	if interval == 0 {
+		interval = time.Minute
+	}
+
+	taskOpts, err := getTaskInfo(t)
+	if err != nil {
+		return nil, errors.Wrap(err, "incomplete task metadata")
+	}
+
+	if len(collectors) == 0 {
+		return nil, errors.New("must provide at least one MetricCollector")
+	}
+
+	bi, err := c.GetBuildloggerInfo(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting cedar dial options")
+	}
+	dialOpts := timber.DialCedarOptions{
+		BaseAddress: bi.BaseURL,
+		RPCPort:     bi.RPCPort,
+		Username:    bi.Username,
+		Password:    bi.Password,
+		APIKey:      bi.APIKey,
+		Retries:     10,
+	}
+	connOpts := system_metrics.ConnectionOptions{
+		DialOpts: dialOpts,
+	}
+
+	return &SystemMetricsCollector{
+		logger:     logger,
+		interval:   interval,
+		collectors: collectors,
+		taskOpts:   taskOpts,
+		connOpts:   connOpts,
+	}, nil
+}
+
+func (s *SystemMetricsCollector) Start(ctx context.Context) {
+	catcher := grip.NewBasicCatcher()
+	defer func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.errored = true
+		err := catcher.Resolve()
+		s.logger.System().Error(message.WrapError(err, message.Fields{
+			"message":  "error setting up system metrics collection",
+			"id":       s.id,
+			"connOpts": s.connOpts,
+			"taskOpts": *s.taskOpts,
+		}))
+	}()
+
+	var err error
+	s.client, err = system_metrics.NewSystemMetricsClient(ctx, s.connOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	innerCtx := context.Background()
+	innerCtx, s.innerCancel = context.WithCancel(innerCtx)
+
+	for _, collector := range s.collectors {
+		s.wg.Add(1)
+		stream, err := s.client.StreamSystemMetrics(innerCtx, s.id, collector.Name(), system_metrics.StreamOpts{})
+		if err != nil {
+			catcher.Add(err)
+			return
+		}
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
+		err = ftdcCollector.SetMetadata(collector.Metadata())
+		if err != nil {
+			catcher.Add(err)
+			return
+		}
+		go s.timedCollect(innerCtx, collector, stream, ftdcCollector)
+	}
+	go s.closeOnCancel(ctx)
+}
+
+func (s *SystemMetricsCollector) timedCollect(ctx context.Context, mc MetricCollector, stream *system_metrics.SystemMetricsWriteCloser, collector ftdc.Collector) {
+	defer s.wg.Done()
+
+	catcher := grip.NewBasicCatcher()
+	defer func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.errored = true
+		err := catcher.Resolve()
+		s.logger.System().Error(message.WrapError(err, message.Fields{
```

Comment:
We should handle the case where if the metrics collector is closed, we should not log to the system logger, because it could occur after `Communicator.EndTask()` is called, which invalidates the logger. The only option in that case is to use something like `grip.Error()` or `grip.Critical()` (which will log to Splunk).